Stm32 g4 variant

ABOSTM opened the pull request:
**Summary**

<!-- Summary of the PR -->
Stm32 g4 variant

fpistm committed:
[variant] Group each variant per STM32 series

As the number of variants continues to grow and in order to
ease navigation, they have been grouped in sub-folders
per STM32 series.

Moreover will ease generic variants management.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
[PIO] Fix variants path changes

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
[ThunderPack] Split variant per series

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Introduce all alternative pins per peripheral

- Alternate pins (PY_n_ALTx)
- Dual pad pins (PY_n_C)
- Remap pins (PY_n_R)

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Extend Arduino pin naming to the max possibilities

- 192 digital pins: GPIO port (A to K + Z) * 16 pins
As Ax is defined as an uint8_t the max number is now:
256 - 192 = 64

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Clean up old way to define Ax pins definition

NUM_ANALOG_FIRST is no more supported.
Only analogInputPin array is.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Rename PinMap_PWM to PinMap_TIM

Define PinMap_PWM as PinMap_TIM for backward compatibility as
several libraries use it. They should be updated to use the new
name.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Remap pin management

Fix #1180

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add Alternate pins (ALTx) management

The PinName mentioned Px_y_ALTz (described in the PeripheralPins.c/PinNames.h files)
are alternative possibilities which use other HW peripheral instances.

To access those alternative PinName name using a pin number can be performed like this:
- PYn_ALTx
- Ax | ALTx
- Dx | ALTx
- x | ALTx

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Change Ax constants to an enum

to avoid conflicts with other (library) code
(i.e. local variables or class attributes named Ax break).

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Definition and comments improvements for pins management

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Rework analog pin management

Use mask to ease analog pin management and ease alternative
analog pins introduction.
Fix some issues.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Clean up generic HAL conf file

Enable IPCC by default
Sorted by name

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Clean up Vddio2 Independent I/Os supply setting

It was already present for STM32L4.
Test the register definition is enough.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

ABOSTM commented:
@fpistm Review requested for last commit:
    [variant] Add Generic G4 and Nucleo G4

fpistm commented on README.md:
```diff
@@ -240,6 +241,14 @@ User can add a STM32 based board following this [wiki](https://github.com/stm32d
+
+| Status | Device(s) | Name | Release | Notes |
+| :----: | :-------: | ---- | :-----: | :---- |
+| :yellow_heart: | STM32G474RE | Generic Board | **2.0.0** |  |
```
The generic provide also support of STM32G474RB and STM32G474RC

fpistm commented on README.md:
```diff
@@ -240,6 +241,14 @@ User can add a STM32 based board following this [wiki](https://github.com/stm32d
+| Status | Device(s) | Name | Release | Notes |
+| :----: | :-------: | ---- | :-----: | :---- |
+| :yellow_heart: | STM32G474RE | Generic Board | **2.0.0** |  |
+| :yellow_heart: | STM32G431RB | Generic Board | **2.0.0** |  |
```
Ditto 6 and 8

fpistm commented on README.md:
```diff
@@ -240,6 +241,14 @@ User can add a STM32 based board following this [wiki](https://github.com/stm32d
+| :----: | :-------: | ---- | :-----: | :---- |
+| :yellow_heart: | STM32G474RE | Generic Board | **2.0.0** |  |
+| :yellow_heart: | STM32G431RB | Generic Board | **2.0.0** |  |
+| :yellow_heart: | STM32G431KB | Generic Board | **2.0.0** |  |
```
ditto 6 and 8

fpistm commented on boards.txt:
```diff
@@ -351,7 +351,7 @@ Nucleo_64.menu.pnum.NUCLEO_G431RB.build.flags.fp=-mfpu=fpv4-sp-d16 -mfloat-abi=h
 Nucleo_64.menu.pnum.NUCLEO_G431RB.build.series=STM32G4xx
 Nucleo_64.menu.pnum.NUCLEO_G431RB.build.product_line=STM32G431xx
-Nucleo_64.menu.pnum.NUCLEO_G431RB.build.variant=STM32G4xx/NUCLEO_G431RB
+Nucleo_64.menu.pnum.NUCLEO_G431RB.build.variant=STM32G4xx/Generic_G431RB
```
Generic_G431Rx

fpistm commented on boards.txt:
```diff
@@ -364,7 +364,7 @@ Nucleo_64.menu.pnum.NUCLEO_G474RE.build.flags.fp=-mfpu=fpv4-sp-d16 -mfloat-abi=h
 Nucleo_64.menu.pnum.NUCLEO_G474RE.build.series=STM32G4xx
 Nucleo_64.menu.pnum.NUCLEO_G474RE.build.product_line=STM32G474xx
-Nucleo_64.menu.pnum.NUCLEO_G474RE.build.variant=STM32G4xx/NUCLEO_G474RE
+Nucleo_64.menu.pnum.NUCLEO_G474RE.build.variant=STM32G4xx/Generic_G474RE
```
Generic_G474Rx

fpistm commented on boards.txt:
```diff
@@ -537,6 +537,19 @@ Nucleo_32.menu.pnum.NUCLEO_G031K8.build.variant=STM32G0xx/Generic_G031Kx
+Nucleo_32.menu.pnum.NUCLEO_G431KB.build.board=NUCLEO_G431KB
+Nucleo_32.menu.pnum.NUCLEO_G431KB.build.series=STM32G4xx
+Nucleo_32.menu.pnum.NUCLEO_G431KB.build.product_line=STM32G431xx
+Nucleo_32.menu.pnum.NUCLEO_G431KB.build.variant=STM32G4xx/Generic_G431KB
```
Generic_G431Kx

fpistm commented on boards.txt:
```diff
@@ -1976,6 +1976,60 @@ GenG0.menu.upload_method.dfuMethod.upload.protocol=2
+GenG4.build.series=STM32G4xx
+GenG4.build.cmsis_lib_gcc=arm_cortexM4lf_math
+
+# Generic G431KB board
+GenG4.menu.pnum.GENERIC_G431KB=Generic G431KB
+GenG4.menu.pnum.GENERIC_G431KB.upload.maximum_size=131072
+GenG4.menu.pnum.GENERIC_G431KB.upload.maximum_data_size=32768
+GenG4.menu.pnum.GENERIC_G431KB.build.board=GENERIC_G431KB
+GenG4.menu.pnum.GENERIC_G431KB.build.product_line=STM32G431xx
+GenG4.menu.pnum.GENERIC_G431KB.build.variant=STM32G4xx/Generic_G431KB
+
+# Generic G431RB board
+GenG4.menu.pnum.GENERIC_G431RB=Generic G431RB
+GenG4.menu.pnum.GENERIC_G431RB.upload.maximum_size=131072
+GenG4.menu.pnum.GENERIC_G431RB.upload.maximum_data_size=32768
+GenG4.menu.pnum.GENERIC_G431RB.build.board=GENERIC_G431RB
+GenG4.menu.pnum.GENERIC_G431RB.build.product_line=STM32G431xx
+GenG4.menu.pnum.GENERIC_G431RB.build.variant=STM32G4xx/Generic_G431RB
+
+# Generic G474RE board
+GenG4.menu.pnum.GENERIC_G474RE=Generic G474RE
+GenG4.menu.pnum.GENERIC_G474RE.upload.maximum_size=524288
+GenG4.menu.pnum.GENERIC_G474RE.upload.maximum_data_size=131072
+GenG4.menu.pnum.GENERIC_G474RE.build.board=GENERIC_G474RE
+GenG4.menu.pnum.GENERIC_G474RE.build.product_line=STM32G474xx
+GenG4.menu.pnum.GENERIC_G474RE.build.variant=STM32G4xx/Generic_G474RE
```
It misses some generic variant for each G4

fpistm commented on boards.txt:
```diff
@@ -2038,7 +2092,25 @@ GenH7.menu.upload_method.dfuMethod.upload.options=-g
 
 ################################################################################
-# Generic L0
+################################
+# Generic G0
+GenG0.name=Generic STM32G0 series
+
+GenG0.build.core=arduino
+GenG0.build.board=GenG0
+GenG0.build.mcu=cortex-m0plus
+GenG0.build.series=STM32G0xx
+GenG0.build.cmsis_lib_gcc=arm_cortexM0l_math
+GenG0.build.extra_flags=-D{build.product_line} {build.enable_usb} {build.xSerial} -D__CORTEX_SC=0
+
+# Generic G030K6
+GenG0.menu.pnum.GENERIC_G030K6=Generic G030K6
+GenG0.menu.pnum.GENERIC_G030K6.upload.maximum_size=32768
+GenG0.menu.pnum.GENERIC_G030K6.upload.maximum_data_size=8192
+GenG0.menu.pnum.GENERIC_G030K6.build.board=GENERIC_G030K6
+GenG0.menu.pnum.GENERIC_G030K6.build.product_line=STM32G030xx
+GenG0.menu.pnum.GENERIC_G030K6.build.variant=STM32G0xx/Generic_G030Kx
+
```
??

fpistm commented on variants/STM32G4xx/Generic_G431KB/variant.cpp:
```diff
@@ -0,0 +1,198 @@
+ *
+ *******************************************************************************
+ */
+#if defined(ARDUINO_GENERIC_G431KB)
```
#if defined(ARDUINO_GENERIC_G431K6)  || defined(ARDUINO_GENERIC_G431K8)  || defined(ARDUINO_GENERIC_G431KB)

fpistm commented on variants/STM32G4xx/Generic_G431KB/variant.cpp:
```diff
@@ -0,0 +1,198 @@
+/*            PLL (clocked by HSE) used as System clock source                */
+/******************************************************************************/
+static bool SetSysClock_PLL_HSE(bool bypass)
+
+
```
to remove

fpistm commented on variants/STM32G4xx/Generic_G431KB/variant.h:
```diff
@@ -0,0 +1,184 @@
+ *
+ *******************************************************************************
+ */
+#if !defined(ARDUINO_GENERIC_G431KB)
```
Ditto missing some defined

fpistm commented on variants/STM32G4xx/Generic_G431KB/variant_NUCLEO_G431KB.cpp:
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
2019-2021

fpistm commented on variants/STM32G4xx/Generic_G431KB/variant_NUCLEO_G431KB.h:
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
2019-2021

fpistm commented on variants/STM32G4xx/Generic_G431RB/variant.cpp:
```diff
@@ -0,0 +1,236 @@
+ *
+ *******************************************************************************
+ */
+#if defined(ARDUINO_GENERIC_G431RB)
```
ditto

fpistm commented on variants/STM32G4xx/Generic_G431RB/variant.cpp:
```diff
@@ -0,0 +1,236 @@
+WEAK void SystemClock_Config(void)
+{
+  /*
+   * HSE_VALUE defined in the stm32g0xx_hal_conf_default.h is used.
```
stm32g4xx_hal_conf_default.h

fpistm commented on variants/STM32G4xx/Generic_G431RB/variant.h:
```diff
@@ -0,0 +1,231 @@
+ *
+ *******************************************************************************
+ */
+#if !defined(ARDUINO_GENERIC_G431RB)
```
ditto

fpistm commented on variants/STM32G4xx/Generic_G431RB/variant_NUCLEO_G431RB.cpp:
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
2019-2021

fpistm commented on variants/STM32G4xx/Generic_G431RB/variant_NUCLEO_G431RB.h:
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
2019-2021

fpistm commented on variants/STM32G4xx/Generic_G474RE/variant.cpp:
```diff
@@ -0,0 +1,239 @@
+ *
+ *******************************************************************************
+ */
+#if defined(ARDUINO_GENERIC_G474RE)
```
ditto

fpistm commented on variants/STM32G4xx/Generic_G474RE/variant.h:
```diff
@@ -0,0 +1,238 @@
+ *
+ *******************************************************************************
+ */
+#if !defined(ARDUINO_GENERIC_G474RE)
```
ditto

fpistm commented on variants/STM32G4xx/Generic_G474RE/variant_NUCLEO_G474RE.cpp:
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
2019-2021

fpistm commented on variants/STM32G4xx/Generic_G474RE/variant_NUCLEO_G474RE.h:
```diff
@@ -1,6 +1,6 @@
 /*
  *******************************************************************************
- * Copyright (c) 2019, STMicroelectronics
+ * Copyright (c) 2021, STMicroelectronics
```
2019-2021

fpistm committed:
Move genpinmap from Arduino_Tools repository

Renamed stm32variant.py

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
[stm32variant] Update to generate in the core

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
[stm32variant] Use Mcu description

This allows to harden the code as some mcu file name
is not enough to define the family, ex:
STM32GBK1CBTx is a STM32G4

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
[stm2variant] Add board_entry.txt

This will help to add generic board entry.
Note:
upload.maximum_size and product_line have to be verified
and changed if needed.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
[stm32variant] Add option to generate only one family

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Allow use of a custom peripheral pins file

Syntax in the board.txt:
xxx.build.peripheral_pins=-DCUSTOM_PERIPHERAL_PINS

Example to use a custom peripheral pins file for Nucleo_L476RG:
    Nucleo_64.menu.pnum.NUCLEO_L476RG.build.peripheral_pins=-DCUSTOM_PERIPHERAL_PINS
Then add a file with ".c" extension like "PeripheralPins_<board>.c" in the
variantfolder.

At sketch folder:
CUSTOM_PERIPHERAL_PINS could now be overridden using the build_opt.h, ex:
-DCUSTOM_PERIPHERAL_PINS=\"PeripheralPins_NUCLEO_L476RG.c\"

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
[stm32variant] Add generic_clock.c generation

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

ttimasdf committed:
Fix HAL not initialized on F105x series

ABOSTM committed:
HardwareSerial: don't call STM32 HAL for every bytes to transfer

Instead of calling HAL for each and every bytes,
try to call it only once for a string.
Several call may be needed to manage internal circular buffer
Fixes #1309

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

fpistm committed:
[stm32variant] Aggregate generated variant if they are the same

To avoid having duplicated variants with all the same files.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

ABOSTM committed:
Enable PWR clock systematically when available

This avoid to miss it in system clock configuration,
as it is not systematically done by cubeMX

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM committed:
spi: wait for end of transfer before returning from spi_transfer()

This guarantee to application that transfer is finished,
and avoid for example that application releases NSS pin
while transfer is still ongoing.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

fpistm committed:
[stm32variant] Comment option to generate from one mcu file

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

ABOSTM committed:
Wire examples: rename old *.pde files to *.ino

Fixes #1319

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

fpistm committed:
[CI] arduino-builder is replaced by arduino-cli

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Define USE_FULL_LL_DRIVER at top level

More and more HAL relies on LL drivers.
So enable the full LL usage to ease access.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Merge pull request #1321 from fpistm/config_cleanup

Review HAL/LL default configuration

fpistm committed:
[stm32variant] Change generic variant generation

- variant.* files are renamed variant_generic.*
Generic variant header file will be included thanks
a single variant.h using a variable defined in boards.txt
- Clean template to remove useless extern "C"
- Include guards replaced by #pragma once directive

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

ABOSTM commented:
@fpistm ,
Rebase on latest automatic generic variant generation.
Fixes CI issue (Astyle + compilation issue)
Take your remarks into account

fpistm committed:
Update variant header management

By default the variant.h will include variant_generic.h file.
*.build.variant_h=variant_{build.board}.h can be added in
the boards.txt to define the dedicated header to a board.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32F0xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32F1xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32F2xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32F3xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32F4xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32F7xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32G0xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32G4xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32H7xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32L0xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32L1xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32L4xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32L5xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32MP1xx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32WBxx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
Add all generated STM32WLxx generic variant files

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
[PIO] Disable PIO build

Fix variant header changes does not work aven if `VARIANT_H`
is added to the 'extra_flags'.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
[variant] Add Generic L4xx

All specific L4 variants moved to the generic variant.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

ABOSTM committed:
[variant] Add Generic G4xx

All specific G4 variants moved to the generic variant.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM commented:
ABOSTM fork. Close this PR
